### PR TITLE
add config option for delimiter

### PR DESCRIPTION
**Pull request by luigi-calori**
This fixes issue #1003 that happen on old module systems that do not support --delim

This fix should leave things as they are, it just honor a line in module.yaml file like:
    delim_format: ' '

For example I use the following to produce spack module files that are usable on our clusters:

**etc/spack/module.yaml**

```yaml
modules:
  tcl:
    naming_scheme: '{name}/{version}-{compiler.name}-{compiler.version}'
    delim_format: ' '
    all:

      autoload: 'direct'
```

### CODE REVIEW

**Comment by citibeth** [comment]
Thanks, this worked for me!

@trws Can you add documentation to this PR? I would create a section in the manual on "old systems", and the various workarounds that are needed on them. Explain how to create a modules.yaml file, etc.

@tgamblin Can we get this merged? The PR looks good to me, and it shouldn't have any adverse effects on people who don't need it.

**Comment by alalazo** [comment]
@citibeth I really wouldn't like to have this merged in `develop`, as the feature : 
- breaks `lua` paths to support a 10 years old technology
- it can be solved by `spack install lmod` without breaking anything

not to mention that 3 lines become 30 with this implementation

**Comment by tgamblin** [comment]
@alalazo @luigi-calori: is the lua breakage because `deilm_format` is set for _all_ modules?

Since we're generating tcl code for `tcl` modules here, could we not just emulate `--delim` ourselves?  For example, instead of using `prepend-path` with `--delim`, could you just generate a `setenv` that uses the old variable value via `$::env(VAR)`?  I am not a TCL module expert so I am not sure how well that would work.

**Comment by alalazo** [comment]
@tgamblin not really, because when you unload the module it will unset the variable, not remove just the path it appended

**Comment by tgamblin** [comment]
@alalazo: it doesn't remember the old value of the variable? Bleh.

Possible better alternative: why don't we just run `modulecmd -V` and omit `--delim` if the version is not recent enough?  What version was `--delim` added in?  I get this on my system:

```
$ modulecmd -V
VERSION=3.2.10
DATE=2012-12-21

AUTOLOADPATH=undef
BASEPREFIX="/usr/share"
BEGINENV=99
CACHE_AVAIL=undef
DEF_COLLATE_BY_NUMBER=undef
DOT_EXT=""
EVAL_ALIAS=1
HAS_BOURNE_FUNCS=1
HAS_BOURNE_ALIAS=1
HAS_TCLXLIBS=undef
HAS_X11LIBS=1
LMSPLIT_SIZE=1000
MODULEPATH="/etc/modulefiles"
MODULES_INIT_DIR="/usr/share/Modules/init"
PREFIX="/usr/share/Modules"
TCL_VERSION="8.5"
TCL_PATCH_LEVEL="8.5.7"
TMP_DIR="/tmp"
USE_FREE=undef
VERSION_MAGIC=1
VERSIONPATH=undef
WANTS_VERSIONING=0
WITH_DEBUG_INFO=undef
```

@luigi-calori: does `modulecmd -V` work on your system?  If not, does it at least return an error status we could check (and omit `--delim` in that case too)?

I'd rather avoid adding additional configuration complexity if we can just autodetect capabilities like this.

**Comment by luigi-calori** [comment]
@alalazo, @tgamblin  I did not know that this was breaking compatibility: I tried to be as compatible and unobtrusive as possible: I' ve added code just for Tcl modules and it should behave as before when no override is present in the yaml file.

If you can give me some hints about which are the problem with lua, I would try to correct.

Regarding spack install lmod, as I' ve tried to explain, in the long run it could be a solution but, in order to "sell" Spack adoption in our environment, It would really help if the module produced could be compatible with the module system currently installed.

Regarding length of implementation, any suggestion more than welcome, I' m far frob being a python guru ;-)
I have implemented thinking of possible extension: I would like to make  module templates more configurable like  #1114, that was the reason for introducing a class to help safe pattern subst.

If you have any better coding patterns to suggest, let me know so I can improve the code.

**Comment by tgamblin** [comment]
@luigi-calori: can you paste the output of `modulecmd -V` on your system?

**Comment by luigi-calori** [comment]
Maybe I miss something:  I have no modulecmd command
our modules use module bash / csh functions .
The version reported by module "command" is:
Modules Release Tcl 3.1.6

I thought it would be good to add more reconfigurability to module code generation ( for example to allow inserting custom tcl code to collect module usage, as it currently done by hand at our site)
If this is not the way to go, the delimiter problem could be resolved by inspecting the module version, not sure where to test and which robust test could be run.
I can try to test python code to retrieve module version

**Comment by tgamblin** [comment]
@luigi-calori: I think in this instance, we don't really need configuration because if you have `--delim`, you know what to do with it, and if you don't have `--delim` you're forced to just omit it.  The configuration just introduces a place for the user to make an error.

I *believe* most module implementations have `modulecmd` laying around somewhere.  Ours is just already in the path:

```bash
$ type module
module is a function
module () 
{ 
    eval `/usr/bin/modulecmd bash $*`
}
```

How'd you get the module version?  It looks like `module -V` works fine too.  Does your output look like this?

```bash 
$ module -V
VERSION=3.2.10
DATE=2012-12-21

#... more stuff ...
```

**Comment by luigi-calori** [comment]
I agree that in this case user configuration is not the best solution. Automatic inspection would be better.

On my system:

$ type module
module is a function
module () 
{ 
    eval `$TCLSH /pico/prod/environment/module/3.1.6/none/modulecmd.tcl bash $*`
}

$ module -V
ERROR: command '-V' not recognized
Modules Release Tcl 3.1.6 ($RCSfile: modulecmd.tcl,v $ $Revision: 1.112 $)
	Copyright GNU GPL v2 1991
Usage: module [ switches ] [ command ]
#usage help.....

Any hint on where and how  add introspection on module?

**Comment by citibeth** [comment]
Sorry I'm not up on the specifics here.  Is this a problem with an old
`modules` or old `tcl`?  Are there any docs online about the `--delim`
feature and how/when/why it was added?

> Any hint on where and how add introspection on module?

I would suggest we introspect by writing a short script that succeeds if
the system supports `--delim` and fails if it does not; this is how
Autotools does it, rather than having to "know" which versions of a system
do/do not support the desired feature.  We then wrap it  all up into a
Python function for use within Spack.

We should consider caching the results of this, and other
system-configuration introspection, between Spack calls.  As the number of
things we introspect grows, it could weight down on startup time for
Spack.  Again, this is like how Autotools does it: the `configure`
(introspection) step is cached, it is not run every time you type `make`.

@citibeth <https://github.com/citibeth> I really wouldn't like to have this
> merged in develop, as the feature :
>
>    - breaks lua paths to support a 10 years old technology
>
> Unfortunately, the supercomputer I use comes with 10-year-old technology
installed.  This seems to be common.


> It would really help if the module produced could be compatible with the
> module system currently installed.


I totally agree.  Our system already has a bunch of hand-built modules,
some of which I use with Spack (for example; to avoid having to build my
own compiler, `git`, etc).  When I type `module avail`, I get all the
hand-built modules, plus all the Spack-built modules.  Having Spack use a
separate module system is asking for pain, even if the two are
theoretically "compatible."  Other considerations:

1. As Luigi mentioned, getting Spack adopted will be easier if I don't tell
sysadmins or other users they have to change modules system.

2. I'm not sold on `lmod`.  Its main advantage over classic modules seems
to be recursive module loading.  But I haven't found that to be a very
useful feature with Spack.  And in order to get it working, you must
install Lua, luarocks, two Lua packages --- and then `lmod`.  Sure, Spack
can automate that, but it sure was a pain when I had to do it to get
EasyBuild running.

@tgamblin <https://github.com/tgamblin> not really, because when you unload
> the module it will unset the variable, not remove just the path it appended


Module unloading is not a very useful feature, IMHO.  I set up an
environment with a bunch of `module load` commands that get me what I need
for a particular project.  I tinker with my environment setup script until
it loads what I need; then I stop messing with it.  If the script is not
right, I log out and start a new shell, allowing me to re-do my loads.

In other words... `module purge` is far more useful than `module unload`.

>
>    - it can be solved by spack install lmod without breaking anything
>
> Yes, Spack can be bootstrapped (I'm writing up stuff on this issue).
BUT... the magic of doing a `git clone spack` and then away you go is lost
if you have to go through a long bootstrap process.  Bootstrapping is also
problematic because it's non-obvious and different on each system.  Walking
up to a new system, you don't really know which parts of it are obsolete
and will need to be replaced by Spack-installed stuff.  You figure that out
as you try to build stuff with Spack and things don't work; and hopefully,
you're able to recognize the problems as bootstrap issues and build the
correct piece.

Bootstrapping is the most laborious, non-automated part of the Spack
process.    Each piece you bootstrap, you have to modify Spack config files
or your `.bashrc`, then go back to `spack install` again.  But once you
finish bootstrapping, you can say `spack install` and have 100 packages get
built with one command.

In short... I see bootstrapping as the biggest current barrier to usability
for Spack.  We should avoid the need to bootstrap whenever possible --- and
it seems to be possible here.

**Comment by alalazo** [comment]
@citibeth Hold on, I am not sure I get your points... I was not suggesting any kind of bootstrapping or hierarchical module layout. 

As far as I understood the use case is : 
- @luigi-calori wants to showcase `spack` providing installed software to users via modules
- he works on an old cluster where modules don't support `--delim`

Proposed solution : 

1. `spack install lmod` (or `environment-modules`)
2. prepare a two lines file that sources the updated module command and adds (as in `module use`) the `tcl` module directory
3. tell the user to source that file to get access to the new software

Maybe I am missing something, but that seemed to me a simple solution if he needs to install alongside system modules...

**Comment by luigi-calori** [comment]
@alalazo  3) tell the user to source that file to get access to the new software

That was exactly what I would like to avoid:   It seems to me that apart from delimiter stuff, modulefiles produced by Spack are "compatible" with my other modulefiles and, if configured with:

naming_scheme: '{name}/{version}-{compiler.name}-{compiler.version}'

they are somewhat compatible with our naming scheme, so that if I ask the user just to do
module use <spack_module_folder>

Then all the modules produced by spack are available and override previously installed modules, so

This way, I can override previously existing modules with spack generated ones

**Comment by alalazo** [comment]
@luigi-calori you are already asking the users to execute a command. What's the difference between : 
```
module use <spack-dir>
```
and
```
source <file>
```
?

**Comment by trws** [comment]
`--delim` was added in modules version 3.2.3 back in 2006 according to their [release notes](https://sourceforge.net/projects/modules/files/Modules/modules-3.2.8/NEWS/view), this is where the 10 year number came from, if you had 10-year-old technology everything would work.  The version being supported here is from 2002.  Am I the only one deeply depressed by having to support broken software from 14 years ago because of lazy sysadmins?  I realize HPC has this problem all over the place, we're switching to gcc 4.9.2 now that most distros are using 5.2 by default for example, but this is really out of control.

Regardless, there are two simple changes we can make:

1.  use `--delim` only when the delimiter is not a colon, that's how I meant to make it work in the first place but doesn't seem to be how I actually implemented it.
2.  when `--delim` is not available we compose the complete environment variable as part of the package and just jam it in there.

The second is sub-optimal, especially because it means that lua modules or others that depend on the values from other modules will end up with huge paths in them, but this combination would solve the issue portably. Of course it does leave the issue that the variables get unset rather than reset back to their previous value, but there's little we can do about that.

**Comment by glennpj** [comment]
> Of course it does leave the issue that the variables get unset rather than reset back to their previous value, but there's little we can do about that.

If this is what you are seeing then something is wrong, or maybe it is due to the very old modules system. Upon `module unload` the variables should get reset to the value prior to the `module load`, assuming it had a value. That is the way it works at my site but the oldest version of environment modules we have is 3.2.7, and that system is about to be retired.

**Comment by luigi-calori** [comment]
@alalazo 
module use is something that can easily be integrated in our module profiles and I think it could be
more easily accepted in production as it does not change the way existing modules are parsed.

As Spack modules use hash, they do not clash with system ones and the command:

module use --append <spack_module_folder> 

add spack modules with lower priority, so can probably be safely added even in production as it should not change behaviour of existing modules.

**Comment by trws** [comment]
The real problem isn't that, it's that without the option there's no way to load modules that need to append to fields with non-colon delimiters without overriding previous ones. An example would be loading Lua-posix, getting a LUA_PATH like "<Lua path>;<Lua posix path>", then loading lua-regex and ending up with "<Lua path>;<Lua regex path>" instead of "<Lua path>;<Lua posix path>;<Lua regex path>".

It's only a problem for things that don't use colon as a delimiter, but for those it's pretty serious.





On August 24, 2016 at 7:43:08 AM PDT, Glenn Johnson <notifications@github.com> wrote:

Of course it does leave the issue that the variables get unset rather than reset back to their previous value, but there's little we can do about that.

If this is what you are seeing then something is wrong, or maybe it is due to the very old modules system. Upon module unload the variables should get reset to the value prior to the module load, assuming it had a value. That is the way it works at my site but the oldest version of environment modules we have is 3.2.7, and that system is about to be retired.

-

**Comment by tgamblin** [comment]
@citbeth: 
> 1. As Luigi mentioned, getting Spack adopted will be easier if I don't tell
sysadmins or other users they have to change modules system.

Yes.

> 2. I'm not sold on `lmod`.  Its main advantage over classic modules seems
to be recursive module loading. 

The main advantage is that it's actually an active, maintained project.  Then hierarchical modules, which make the view of what's installed simpler for users.  TCL modules have recursive loading already -- that's not a unique Lmod feature.  @alalazo's point is that Lmod is backward-compatible with TCL modules.  I'm not as convinced by that point b/c I think we should make things easy for users, and their systems might not have Lmod natively.  Ideally Spack would work out of the box.  Also if this makes Spack an easier sell at CINECA, then I think that's a good thing and could result i them using something like Lmod eventually.

Basically, I agree users shouldn't have to install Lmod, but I would like Lmod to take over the modules space.

> 3. Module unloading is not a very useful feature, IMHO.  I set up an
environment with a bunch of `module load` commands that get me what I need
for a particular project.

Lots of people use module unloading.  I don't think it would be a complete solution if we didn't support it.  Also, `module purge` is basically broken on some systems (On Cray you can't module purge and easily reconstruct a vanilla environment -- there are more steps 😱)

> In short... I see bootstrapping as the biggest current barrier to usability
for Spack.  We should avoid the need to bootstrap whenever possible --- and
it seems to be possible here.

Yes!  I want easy bootstrapping.

**Comment by tgamblin** [comment]
@alalazo @luigi-calori @citibeth:

How are you with @trws's option 1?

> use --delim only when the delimiter is not a colon, that's how I meant to make it work in the first place but doesn't seem to be how I actually implemented it.

This would:

1. Eliminate the need to do any auto-detection of `modules` version.
2. Fix packages that do use `:` as an identifier with pre-3.2 `modules` versions.
3. Cause all modules to work with modern `modules` versions

Lua packages would still be broken with older modules versions, but they'd be broken anyway because the paths would use the wrong delimiter.  I think it's reasonable to require a new version of `modules` to get proper Lua support.  I also think it's unreasonable for us to jump through hoops to make old versions of `modules` do things that have been fixed for 10 years.

Thoughts?

**Comment by alalazo** [comment]
@tgamblin fully agree with the last post

**Comment by citibeth** [comment]
>
> How are you with @trws <https://github.com/trws>'s option 1?
> use --delim only when the delimiter is not a colon, that's how I meant to
> make it work in the first place but doesn't seem to be how I actually
> implemented it.


I like it.  Spack will still break for people who use Lua on ancient
systems.  But that's better than the situation now, where it breaks for
everyone on ancient systems.

BTW... this solution, without further fixes for non-colon delimiters, will
prevent me from using Lua or LMod on our supercomputer.  I'm OK with that,
but just saying...  And I agree with Tom, Spack should work with whatever
modules system the sysadmins have decided to use.

**Comment by trws** [comment]
On 24 Aug 2016, at 9:48, Elizabeth Fischer wrote:

> BTW... this solution, without further fixes for non-colon delimiters, 
> will
> prevent me from using Lua or LMod on our supercomputer.  I'm OK with 
> that,
> but just saying...  And I agree with Tom, Spack should work with 
> whatever
> modules system the sysadmins have decided to use.

This is actually the reason for the second point.  If it does that, then 
you could load lmod with spack and it would work.  That said, it would 
leave all of the modules built referring to the old module system 
broken, so there's definitely still an issue there.

**Comment by tgamblin** [comment]
@trws: do you mind doing a PR for 1 (and 2 if you really really want to?)

**Comment by citibeth** [comment]
On Wed, Aug 24, 2016 at 12:54 PM, Tom Scogland <notifications@github.com>
wrote:

>
>
> On 24 Aug 2016, at 9:48, Elizabeth Fischer wrote:
>
> > BTW... this solution, without further fixes for non-colon delimiters,
> > will
> > prevent me from using Lua or LMod on our supercomputer. I'm OK with
> > that,
> > but just saying... And I agree with Tom, Spack should work with
> > whatever
> > modules system the sysadmins have decided to use.
>
> This is actually the reason for the second point. If it does that, then
> you could load lmod with spack and it would work. That said, it would
> leave all of the modules built referring to the old module system
> broken, so there's definitely still an issue there.


The second point makes life worse for users who are happy with the way
things are today: everyone using a version of modules less than 10 years
old.  Plus it will take effort to implement.

If we implement just point 1, we improve some users' lives while not making
things worse for anyone.  And it's less effort to boot.  For the 1% of
users who have ancient systems AND need to use Lua... maybe we suggest that
they bootstrap `lmod` as @alanzo suggested.  I vote that we don't we cross
that bridge unless we actually come to it.

**Comment by tgamblin** [comment]
@trws: I lean towards @citibeth's opinion on option 2.

**Comment by trws** [comment]
I'm working on getting part 1 done.  It's actually surprisingly 
cumbersome, would have thought python would have a conditional format 
operator but apparently not.  Anyway, it shouldn't take long, just 
testing it now to make sure it behaves itself.

**Comment by luigi-calori** [comment]
I' m not sure to have well understood exactly how to find if the delimiter is not a colon ( do you need to parse value at line 534, 535,536  searching for non colon delimiter?) but It seems the simplest option that allow use of old tcl modules without requiring user configuration.
Waiting to test  it ;-)
Nevertheless, I vote for increasing user customization of modules templates

**Comment by tgamblin** [comment]
@luigi-calori: we know what the delimiter is from the python side, e.g. [this line](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/lua/package.py#L105) in the Lua package.  So we can ignore the separator and not translate it to `--delim` in the module file when it is not `:`.

**Comment by tgamblin** [comment]
@luigi-calori: This is resolved by #1606.  I'd be happy to talk about ways to customize modules further -- I just think the `--delim` thing is better solved as discussed here.
